build_modules/lib/src/meta_module.dart

```diff
@@ -0,0 +1,433 @@
+    newModule.directDependencies.removeAll(newModule.sources);
+  }
+
+  return modulesById.values.toList();
+}
+
+// Returns whether [dart] contains a [PartOfDirective].
+bool _isPart(CompilationUnit dart) =>
+    dart.directives.any((directive) => directive is PartOfDirective);
+
+/// Returns whether [dart] looks like an entrypoint file.
+bool _isEntrypoint(CompilationUnit dart) {
+  return dart.declarations.any((node) {
+    return node is FunctionDeclaration &&
+        node.name.name == 'main' &&
+        node.functionExpression.parameters.parameters.length <= 2;
+  });
+}
+
+Future<List<Module>> _computeModules(
+    AssetReader reader, List<AssetId> assets, bool public) async {
+  var dir = _topLevelDir(assets.first.path);
+  if (!assets.every((src) => _topLevelDir(src.path) == dir)) {
+    throw new ArgumentError(
+        'All srcs must live in the same top level directory.');
+  }
+
+  // The set of entry points from `srcAssets` based on `mode`.
+  var entryIds = new Set<AssetId>();
+  // All the `srcAssets` that are part files.
+  var partIds = new Set<AssetId>();
+  // Invalid assets that should be removed from `srcAssets` after this loop.
+  var idsToRemove = <AssetId>[];
+  var parsedAssetsById = <AssetId, CompilationUnit>{};
+  for (var asset in assets) {
+    var content = await reader.readAsString(asset);
+    // Skip errors here, dartdevc gives nicer messages.
+    var parsed = parseCompilationUnit(content,
+        name: asset.path, parseFunctionBodies: false, suppressErrors: true);
+    parsedAssetsById[asset] = parsed;
+
+    // Skip any files which contain a `dart:_` import.
+    if (parsed.directives.any((d) =>
+        d is UriBasedDirective &&
+        d.uri.stringValue.startsWith('dart:_') &&
+        asset.package != 'dart_internal')) {
+      idsToRemove.add(asset);
+      continue;
+    }
+
+    // Short-circuit for part files.
+    if (_isPart(parsed)) {
+      partIds.add(asset);
+      continue;
+    }
+
+    if (public) {
+      if (!asset.path.startsWith('lib/src/')) entryIds.add(asset);
+    } else {
+      if (_isEntrypoint(parsed)) entryIds.add(asset);
+    }
+  }
+
+  var trimedAssets =
+      assets.where((asset) => !idsToRemove.contains(asset)).toList();
+  assets = trimedAssets;
+  // Build the `_AssetNode`s for each asset, skipping part files.
+  var nodesById = <AssetId, _AssetNode>{};
+  var srcAssetIds = assets.map((asset) => asset).toSet();
+  var nonPartAssets = assets.where((asset) => !partIds.contains(asset));
+  for (var asset in nonPartAssets) {
+    var node = new _AssetNode.forParsedUnit(
+        asset, parsedAssetsById[asset], srcAssetIds);
+    nodesById[asset] = node;
+  }
+  var connectedComponents = _stronglyConnectedComponents(nodesById);
+  var modulesById = _createModulesFromComponents(connectedComponents);
+  var modules = _mergeModules(modulesById, entryIds);
+  return modules;
+}
+
+@JsonSerializable()
+class MetaModule extends Object with _$MetaModuleSerializerMixin {
```

Comment:
nit: consider separating this into its own file.

Reply:
I can see moving the computerModule logic over to another file but going to leave it as is for now.